plugins/gitlab-catalog-backend/src/GitlabUserProcessor.ts

```diff
@@ -0,0 +1,148 @@
+  private loadUserPromise: Promise<Map<string, ExpandedUserSchema>> | null =
+    null;
+
+  private async fetchUsers(): Promise<Map<string, ExpandedUserSchema>> {
+    if (!this.gitlab) {
+      return new Map();
+    }
+
+    if (!this.cacheLoaded) {
+      // we use a shared promise to make sure that the load is only
+      // executed once even if `GitlabUserProcessor.postProcessEntity`
+      // is called "concurrently".
+      if (!this.loadUserPromise) {
+        this.loadUserPromise = this.loadUsers();
+      }
+
+      // Wait for the data to load
+      await this.loadUserPromise;
+    }
+
+    return this.userLookup;
+  }
+
+  private async loadUsers(): Promise<Map<string, ExpandedUserSchema>> {
+    this.logger.info('Loading gitlab users');
+
+    let members: ExpandedUserSchema[] = [];
+    try {
+      members = await this.gitlab!.Users.all({
+        perPage: GITLAB_PER_PAGE_LIMIT,
+      });
+    } catch (error) {
+      this.logger.error(`Error loading gitlab users: ${error}`);
+      return this.userLookup;
+    }
+
+    members.forEach(user => {
+      if (user.email) {
+        this.userLookup.set(user.email, user);
+      }
+    });
+
+    this.logger.info(`Loaded ${this.userLookup.size} gitlab users`);
+    this.cacheLoaded = true;
+
+    return this.userLookup;
+  }
+
+  static fromConfig(config: Config, logger: Logger): GitlabUserProcessor[] {
+    const token = config.getOptionalString('gitlabCatalog.token');
+    if (!token) {
+      logger.warn(
+        'No token provided for GitlabUserProcessor, skipping Gitlab user lookup',
+      );
+      return [];
+    }
+    let host = config.getOptionalString('gitlabCatalog.host');
+    if (!host) {
+      logger.info(
+        `No host provided for GitlabUserProcessor, defaulting to ${GITLAB_DEFAULT_HOST}`,
+      );
+      host = GITLAB_DEFAULT_HOST;
+    }
+    return [
+      new GitlabUserProcessor(
+        new Gitlab({
+          host: host,
+          token: token,
+        }),
+        logger,
+      ),
+    ];
+  }
+
+  // @ts-ignore: intended to reference as such by the library
+  constructor(gitlab: Gitlab, logger: Logger) {
+    this.gitlab = gitlab;
+    this.logger = logger;
+    this.userLookup = new Map();
+    this.cacheLoaded = false;
+  }
+
+  getProcessorName(): string {
+    return 'GitlabUserProcessor';
+  }
+
+  async postProcessEntity(
+    entity: Entity,
+    _location: LocationSpec,
+    _emit: CatalogProcessorEmit,
+  ): Promise<Entity> {
+    if (!isUserEntity(entity)) {
+      return entity;
+    }
+
+    const email = entity.spec.profile?.email;
+    if (!email) {
+      return entity;
+    }
+
+    const userLookup = await this.fetchUsers();
+    const gitlabUser = userLookup.get(email);
+    if (!gitlabUser) {
+      return entity;
+    }
+
+    if (!entity.metadata.annotations) {
+      entity.metadata.annotations = {};
+    }
+
+    if (gitlabUser.id) {
+      entity.metadata.annotations[`gitlab.com/user_id`] =
```

Comment:
decided to leave simply as gitlab.com no matter if its self-hosted or not, much easier to reference and id imagine people will only ever have one gitlab instance at a time